pkg/store/labelpb/label.go

```diff
@@ -50,12 +50,34 @@ func ZLabelsToPromLabels(lset []ZLabel) labels.Labels {
 }
 
 // ReAllocZLabelsStrings re-allocates all underlying bytes for string, detaching it from bigger memory pool.
-func ReAllocZLabelsStrings(lset *[]ZLabel) {
+// It also takes a 'cache' argument, which is custom map type labelInterningMap. This map is used to do string
+// interning, i.e. reusing already allocated strings, to make the reallocation method more efficient.
+//
+// Due to the charaterstics of this method described above, it should be used with care and only
+// for immutable label sets. This is primarily intended to be used before labels are written into TSDB
+// which can hold label strings in the memory long term.
+func ReAllocZLabelsStrings(lset *[]ZLabel, cache labelInterningMap) {
 	for j, l := range *lset {
-		// NOTE: This trick converts from string to byte without copy, but copy when creating string.
-		(*lset)[j].Name = string(noAllocBytes(l.Name))
-		(*lset)[j].Value = string(noAllocBytes(l.Value))
+		(*lset)[j].Name = cache.internOrReallocateString(l.Name)
+		(*lset)[j].Value = cache.internOrReallocateString(l.Value)
+	}
+}
+
+// labelInterningMap is type to enable string interning for labels.
+type labelInterningMap map[string]string
+
+// internOrReallocateString is a method that checks the interning map
+// for an existing (cached) string. If it exists, the interned string is
+// returned. If not, a new string is allocated and cached in the map.
+func (l labelInterningMap) internOrReallocateString(s string) string {
+	interned, ok := l[s]
```

Comment:
Do we require a lock here, or is the cache scoped to a single request?

Reply:
For the reasons stated in the PR description, this is only scoped to single request. Probably a good idea to document this in the method / type description as well.

Reply:
Would you mind pointing out what is the exact reason again? I re-read the description but I could not understand what's the problem with a cross-request cache. I also looked at the linked Prometheus issue, but that just says that the way we store labels will change. Is that the reason why we don't want to go beyond a single request?

Reply:
Got it. Yeah a TTL does make sense. Maybe we can start with just resetting the map and then we can add some smart tracking? Doing a reset every 1h should be good enough I think, but we can experiment with it.